src/main/java/io/aiven/kafka/connect/http/sender/OAuth2HttpSender.java

```diff
@@ -19,82 +19,116 @@
+        // This handler allows to request a new access token if a 401 occurs, meaning the session might be expired
+        final HttpResponseHandler handler = (response, remainingRetries) -> {
+            // If the response has a 401 error and we have retries left, we attempt to renew the session
+            if (response.statusCode() == 401 && remainingRetries > 0) {
+                // Update the request builder with the new access token
+                ((OAuth2AuthHttpRequestBuilder) this.httpRequestBuilder).renewAccessToken(requestBuilder);
+                // Retry the call and decrease the retries counter to avoid looping on token renewal
+                this.sendWithRetries(requestBuilder, originHttpResponseHandler, remainingRetries - 1);
             } else {
-                originHttpResponseHandler.onResponse(response);
+                originHttpResponseHandler.onResponse(response, remainingRetries);
             }
         };
-        return super.sendWithRetries(accessTokenAwareRequestBuilder, handler);
+        return super.sendWithRetries(requestBuilder, handler, retries);
     }
 
-    private HttpRequest.Builder loadOrRefreshAccessToken(final HttpRequest.Builder requestBuilder) {
-        if (accessTokenAuthHeader == null) {
-            logger.info("Configure OAuth2 for URI: {} and Client ID: {}",
-                    config.oauth2AccessTokenUri(), config.oauth2ClientId());
+    static class OAuth2AuthHttpRequestBuilder extends DefaultHttpRequestBuilder {
+
+        private static final Logger LOGGER = LoggerFactory.getLogger(OAuth2AuthHttpRequestBuilder.class);
+        private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+        private final HttpSinkConfig config;
+        private final AccessTokenHttpSender accessTokenHttpSender;
+
+        private String accessToken;
+
+        OAuth2AuthHttpRequestBuilder(final HttpSinkConfig config, final AccessTokenHttpSender accessTokenHttpSender) {
+            this.config = config;
+            this.accessTokenHttpSender = accessTokenHttpSender;
+        }
+
+        @Override
+        public Builder build(final HttpSinkConfig config) {
+            return super
+                .build(config)
+                // We need to retrieve an access token first
+                .header(HEADER_AUTHORIZATION, getAccessToken());
+        }
+
+        /**
+         * When expired, reinitialize the current token, request a new one and update the request builder
+         * @param requestBuilder the request builder used to call the protected URI
+         */
+        void renewAccessToken(final HttpRequest.Builder requestBuilder) {
+            this.accessToken = null;
+            requestBuilder.setHeader(HttpRequestBuilder.HEADER_AUTHORIZATION, this.getAccessToken());
+        }
+
+        /**
+         * Retrieves the current access token or requests it if none is defined
+         * @return an access token
+         */
+        private String getAccessToken() {
```

Comment:
nit: maybe a bit more specific?
```suggestion
        private String requestAccessToken() {
```